### PR TITLE
 FEATURE: custom suppression patterns via UI

### PR DESCRIPTION
**Pull request by OsamaSayegh**
This feature adds a user interface to extend the current logs suppression feature which allows you to suppress logs that match regular expression patterns. Previously it you wanted suppress a certain log message you would have to commit a change to your app code and then deploy it. This is still possible if you want to permanently suppress a pattern, but now you can quickly add/remove patterns without deploying your app.

Note: adding a new pattern will NOT remove existing logs that match the new pattern.

Screenshot:

image

### CODE REVIEW

File: lib/logster/base_store.rb [hunk 1]
```diff
@@ -84,6 +84,21 @@ def check_rate_limits(severity)
...
+    end
+
+    # takes a string as `pattern` and removes it from the set `set_name`
+    def remove_pattern(set_name, pattern)
+      not_implemented
+    end
+
+    # returns an array of strings each of which must be convertible to regexp
+    def get_patterns(set_name)
+      not_implemented
```
Comment (SamSaffron) [inline lib/logster/base_store.rb]: I don't mind this pattern, some people complain raising https://ruby-doc.org/core-2.2.0/NotImplementedError.html is a misuse which is true, so this is a good enough clean workaround imo

File: lib/logster/base_store.rb [hunk 2]
```diff
@@ -110,6 +125,11 @@ def report(severity, progname, msg, opts = {})
 
       return if ignore && ignore.any? { |pattern| message =~ pattern }
 
+      if Logster.config.enable_custom_patterns_via_ui
+        custom_ignore = Logster::SuppressionPattern.find_all
```
Comment (SamSaffron) [inline lib/logster/base_store.rb]: is this reaching to redis every time? can it be cached?

File: test/logster/middleware/test_viewer.rb [hunk 3]
```diff
@@ -18,6 +18,7 @@ def setup
 
   def teardown
     Logster.config.subdirectory = nil
+    Logster.store.redis.flushall
```
Comment (SamSaffron) [inline test/logster/middleware/test_viewer.rb]: maybe we should avoid this atomic destruction and instead kill off the parts of the store we know we touched? that keeps our "clean up redis mess" routines honest and working.

File: test/logster/middleware/test_viewer.rb [hunk 4]
```diff
@@ -60,6 +61,159 @@ def test_regex_parse
...
+    response = request.post("/logsie/patterns/suppression.json",
+      params: { pattern: "patternfromuser" }
+    )
+    assert_equal(403, response.status)
+    assert_equal(0, Logster::SuppressionPattern.find_all.size)
+  end
+
+  def test_patterns_endpoint_doesnt_work_with_undefined_set
+    Logster.config.enable_custom_patterns_via_ui = true
+    Logster.store.redis.flushall
```
Comment (SamSaffron) [inline test/logster/middleware/test_viewer.rb]: more flushall, ideally lets avoid these in favor of strategic calls

**Comment by SamSaffron** [comment]
I love this change, lets just make sure we aim for very low performance impact even if custom regexs are defined.

**Comment by SamSaffron** [comment]
note ... cache is going to be hard due to the multi process nature of this, but I guess I am ok with... 

1. default implementation having a 2 second cache in place, just to handle bursts
2. a custom cache provider can use something like the message bus distributed cache, which uses message bus and gives you clean ultra fast expiry with ultra cheap gets

**Comment by OsamaSayegh** [comment]
@SamSaffron I've removed `flushall`s and added cache for suppression patterns. Could you review this please? :)

File: client-app/app/styles/app.css [hunk 5]
```diff
@@ -1,5 +1,5 @@
 body {
-  font-family: "Roboto", Arial, "Liberation Sans", "DejaVu Sans", sans-serif;
```
Comment (SamSaffron) [inline client-app/app/styles/app.css]: hmmm I recall Jeff wanted Roboto, any reason you want it removed?

File: lib/logster/base_store.rb [hunk 6]
```diff
@@ -110,6 +127,13 @@ def report(severity, progname, msg, opts = {})
 
       return if ignore && ignore.any? { |pattern| message =~ pattern }
 
+      if Logster.config.enable_custom_patterns_via_ui || allow_custom_ignore
+        custom_ignore = @patterns_cache.get do
```
Comment (SamSaffron) [inline lib/logster/base_store.rb]: use the word `fetch` here, that matches what hash does.

File: lib/logster/cache.rb [hunk 7]
```diff
@@ -0,0 +1,16 @@
+module Logster
+  class Cache
+    def initialize(age = 2)
+      @age = age
+      @hash = { created_at: Time.now.to_f }
+    end
+
+    def get
+      if !@hash.key?(:data) || @hash[:created_at] + @age < Time.now.to_f
```
Comment (SamSaffron) [inline lib/logster/cache.rb]: I hate pointing at micro optimisations... but its worth noting ... `Process.clock_gettime(Process::CLOCK_MONOTONIC)` this is good for single process clock, its cheaper to ask for this cause you allocate one less object and is slightly more accurate.

File: lib/logster/cache.rb [hunk 8]
```diff
@@ -0,0 +1,16 @@
+module Logster
+  class Cache
```
Comment (SamSaffron) [inline lib/logster/cache.rb]: when we implement one for core we also need a #clear hear. and to make sure #clear is being called when amending stuff via the UX.

File: test/logster/test_redis_store.rb [hunk 9]
```diff
@@ -712,9 +727,32 @@ def test_rate_limits_with_prefix
...
+      assert_equal(1, latest.size)
+      assert_equal("myforest", latest.first.message)
+    end
+  end
+
   private
 
   def reset_redis
-    @store.redis.flushall
+    @store.redis.keys.each do |key|
```
Comment (SamSaffron) [inline test/logster/test_redis_store.rb]: my point here is actually a bit more that this. I would like us to use proper logster apis for clearing stuff. We should never be reaching to redis from the tests here, its an implementation detail. 

```
@store.clear_suppressions! 
@store.clear_logs!
```
or

```
@store.clear_all
```

File: client-app/app/styles/app.css (hunk 5 above)
Comment (OsamaSayegh) [reply client-app/app/styles/app.css]: I made this change for the new settings page which looked really weird with Roboto (it looks the same even with a really low font-weight)
Screenshot_20190318-084113

Compare this with how Roboto looks on desktop:

image


I didn't remove Roboto completely, the "homepage" still looks the same, I just moved Roboto from `body` to specific elements: https://github.com/discourse/logster/pull/82/files/a9aead48dc135cdee6b0000e8897bb8a3dc9ba09#diff-fa37db806cc56aa352948366c4a7a53cR37

Comment (SamSaffron) [reply client-app/app/styles/app.css]: cool, makes sense.

File: test/logster/test_redis_store.rb (hunk 9 above)
Comment (OsamaSayegh) [reply test/logster/test_redis_store.rb]: I've removed direct redis calls in tests as much as I can. There are only 2 places left:

1. https://github.com/discourse/logster/blob/master/test/logster/test_redis_rate_limiter.rb which seems to be fundamentally based on redis and pulling out redis calls from it is going to be very tricky.

2. https://github.com/discourse/logster/blob/5bc5014f66b059becf92508f56214abb7ef2f859/test/logster/test_redis_store.rb#L160-L193 these 2 tests were added when we separated env from main message body. The purpose of these tests is to ensure backward compatibility which requires a couple hacks. When I wrote these tests I added a note that these can be removed in 6-12 months. I can remove them now if you want.

File: lib/logster/redis_store.rb [hunk 10]
```diff
@@ -272,6 +272,12 @@ def clear_all
       @redis.del(env_key)
       @redis.del(grouping_key)
       @redis.del(solved_key)
```
Comment (SamSaffron) [inline lib/logster/redis_store.rb]: tricky tricky .... "clear all" button in the UX should not nuke patterns, so you need this thing to take a param in to say "clear patterns as well"

**Comment by SamSaffron** [comment]
OK once you address my concern... go ahead and merge!

File: test/logster/test_redis_store.rb (hunk 9 above)
Comment (SamSaffron) [reply test/logster/test_redis_store.rb]: OK go ahead and remove the backwards compat test, we can just live with the rate limiting test for now, it is tricky to abstract into an interface.

File: lib/logster/redis_store.rb (hunk 10 above)
Comment (OsamaSayegh) [reply lib/logster/redis_store.rb]: Maybe I'm misunderstanding here, but this method is not called when you hit "clear all" in the UX. Instead [`Store#clear`](https://github.com/discourse/logster/blob/5bc5014f66b059becf92508f56214abb7ef2f859/lib/logster/redis_store.rb#L233) is called:

https://github.com/discourse/logster/blob/5bc5014f66b059becf92508f56214abb7ef2f859/lib/logster/middleware/viewer.rb#L86-L91

I did a quick search and this method is only used in tests.

But I think it's worth adding a test case for this particular case to make sure we don't regress.

**Pull request merged**